internal/notification/smtp_notifier.go

```diff
@@ -1,52 +1,112 @@
 package notification
 
 import (
+	"crypto/tls"
+	"errors"
 	"fmt"
 	"net/smtp"
+	"strings"
 
 	"github.com/authelia/authelia/internal/configuration/schema"
+	"github.com/authelia/authelia/internal/utils"
+	log "github.com/sirupsen/logrus"
 )
 
 // SMTPNotifier a notifier to send emails to SMTP servers.
 type SMTPNotifier struct {
-	address  string
-	sender   string
 	username string
 	password string
+	sender   string
 	host     string
 	port     int
+	secure   bool
+	address  string
 }
 
 // NewSMTPNotifier create an SMTPNotifier targeting a given address.
 func NewSMTPNotifier(configuration schema.SMTPNotifierConfiguration) *SMTPNotifier {
 	return &SMTPNotifier{
+		username: configuration.Username,
+		password: configuration.Password,
+		sender:   configuration.Sender,
 		host:     configuration.Host,
 		port:     configuration.Port,
+		secure:   configuration.Secure,
 		address:  fmt.Sprintf("%s:%d", configuration.Host, configuration.Port),
-		sender:   configuration.Sender,
-		username: configuration.Username,
-		password: configuration.Password,
 	}
 }
 
-func (n *SMTPNotifier) authenticatedSend(recipient string, msg string) error {
-	auth := smtp.PlainAuth("", n.username, n.password, n.host)
-	err := smtp.SendMail(fmt.Sprintf("%s:%d", n.host, n.port), auth, n.sender,
-		[]string{recipient}, []byte(msg))
-	if err != nil {
-		return err
-	}
-	return nil
-}
+// Send send a identity verification link to a user.
+func (n *SMTPNotifier) Send(recipient string, subject string, body string) error {
+	msg := "From: " + n.sender + "\n" +
+		"To: " + recipient + "\n" +
+		"Subject: " + subject + "\n" +
+		"Content-Type: text/html\n" +
+		"MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n" +
+		body
 
-func (n *SMTPNotifier) unauthenticatedSend(recipient string, msg string) error {
-	// Connect to the remote SMTP server.
 	c, err := smtp.Dial(n.address)
 
 	if err != nil {
 		return err
 	}
 
+	// Do StartTLS if available (some servers only provide the auth extnesion after, and encrpytion is preferred)
+	starttls, _ := c.Extension("STARTTLS")
+	if starttls {
+		tlsconfig := &tls.Config{
+			InsecureSkipVerify: !n.secure,
+			ServerName:         n.host,
+		}
+		log.Debugf("SMTP server supports STARTTLS (InsecureSkipVerify: %t, ServerName: %s), attempting", tlsconfig.InsecureSkipVerify, tlsconfig.ServerName)
+		err := c.StartTLS(tlsconfig)
+		if err != nil {
+			return err
+		} else {
+			log.Debug("SMTP STARTTLS completed without error")
+		}
+	} else {
+		log.Debug("SMTP server does not support STARTTLS, skipping")
+	}
+
+	// Attempt AUTH if password is specified only
+	if n.password != "" {
+
```

Comment:
Add a warning when authentication takes place without STARTTLS. This should raise a security red flag to the administrator.
We could print something like "Authentication occurred over an insecure connection. Using STARTTLS is recommended to keep credentials safe."

Reply:
Good idea.